lib/rabbit_feed/event.rb

```diff
@@ -2,42 +2,52 @@ module RabbitFeed
   class Event
     include ActiveModel::Validations
 
-    attr_reader :schema, :payload
-    validates_presence_of :schema, :payload
+    attr_reader :schema, :payload, :metadata
+    validates :metadata, presence: true
+    validates :payload, length: { minimum: 0, allow_nil: false, message: 'can\'t be nil' }
 
-    def initialize schema, payload
-      @schema  = schema
-      @payload = payload
+    def initialize metadata, payload={}, schema=nil
+      @schema   = schema
+      @payload  = payload.with_indifferent_access if payload
+      @metadata = metadata.with_indifferent_access if metadata
       validate!
     end
 
+    def self.from_pre_2_0 metadata_and_payload, schema
+      metadata = {}
+      %w(application name host version environment created_at_utc).each do |field|
+        metadata[field] = metadata_and_payload.delete field
+      end
+      new metadata, metadata_and_payload, schema
+    end
+
     def serialize
       buffer = StringIO.new
       writer = Avro::DataFile::Writer.new buffer, (Avro::IO::DatumWriter.new schema), schema
-      writer << payload
+      writer << { 'metadata' => metadata, 'payload' => payload }
       writer.close
       buffer.string
     end
 
-    def self.deserialize event
+    def self.deserialize serialized_event
       datum_reader = Avro::IO::DatumReader.new
-      reader       = Avro::DataFile::Reader.new (StringIO.new event), datum_reader
-      payload      = nil
+      reader       = Avro::DataFile::Reader.new (StringIO.new serialized_event), datum_reader
+      event_hash   = nil
       reader.each do |datum|
-        payload = datum
+        event_hash = datum
       end
       reader.close
-      Event.new datum_reader.readers_schema, payload
-    end
-
-    def method_missing(method_name, *args, &block)
-      payload[method_name.to_s]
+      if %w(metadata payload).all?{|key| event_hash.has_key? key}
```

Comment:
is not a major but would read better `if version_1?` and then create that method with the content of this if.